Fixes #4544 -- error when attempting to zoom to invalid dates

erikmaarten opened the pull request.

erikmaarten commented:
Might also fix #4425.

mramato commented on Source/Widgets/Timeline/Timeline.js:
```diff
@@ -251,6 +253,11 @@ define([
         }
         //>>includeEnd('debug');
 
+        var timesValid = JulianDate.greaterThanOrEquals(startTime, Iso8601.MINIMUM_VALUE) && JulianDate.lessThanOrEquals(stopTime, Iso8601.MAXIMUM_VALUE);
```
@emackey can confirm but I suspect we would want to throw in this case instead of a no-op, since a developer might call this code directly.  We should then do this check before our own calls to `zoomTo` instead so that we only no-op in those cases.

erikmaarten replied:
Right, so in case someone else calls the `zoomTo` with invalid time arguments we'll just throw and make it clear that the arguments are outside the valid time span. Are there plug-ins or something similar that integrate or could be expected to integrate with the timeline?

mramato replied:
I'm not sure what you mean by plug-ins, by "our own code" I just meant `Viewer.js` (and maybe within the Timeline itself if it calls it).

We already crashed in this case, right?  So the only real change here is being more explicit about the error.

erikmaarten commented:
@mramato I've added validity checks now before our calls to `zoomTo`, and `zoomTo` will throw when called with invalid time arguments. That's what you meant, right?

I also added a static function on JulianDate for this kind of validity check, since it's something that needs to be done in a few different places. Not entirely sure what to do if `Viewer.js` has clock values outside the Iso8601 range and wants to update the timeline, but do we need to take such a case into consideration?

Thanks @mramato!

erikmaarten commented on Source/Core/JulianDate.js:
```diff
@@ -957,6 +959,30 @@ define([
+        }
+        //>>includeEnd('debug');
+
+        var MINIMUM_VALUE = JulianDate.fromIso8601('0000-01-01T00:00:00Z');
```
Not very clean to include this here, since it's available in `Iso8601.js` -- but I can't include it because it'll create a circular dependency.

erikmaarten commented:
After stumbling upon a browser support discussion in another PR I realized I shouldn't use `Array.prototype.every` since it seems to not be available in IE 8.

Ready for review.

erikmaarten commented on Source/Core/JulianDate.js:
```diff
@@ -976,10 +976,15 @@ define([
+
+        for (var i = 0; i < julianDates.length; i++) {
+            var julian = julianDates[i];
+            datesAllValid = datesAllValid && JulianDate.greaterThanOrEquals(julian, MINIMUM_VALUE) && JulianDate.lessThanOrEquals(julian, MAXIMUM_VALUE);
```
Could be more efficient by returning early when it finds an invalid date, but given that I don't currently see a use case for checking that many dates at a time, the only noticeable effect would be a few lines of code more.

erikmaarten commented:
@mramato Could you review this?

mramato commented:
While looking over the code I realized that some assumptions I had regarding the root cause of this issue may have been wrong.  I need to investigate some more and convince myself this is still the right approach.  I'll keep you posted.  Thanks.

mramato commented:
I just opened #4595, sorry for the confusion here.
